java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java

```diff
@@ -1127,4 +1127,24 @@ public void lookup_method_on_defered_type() throws Exception {
     //Lack of resolution when target type is deduced, we should be able to redo the lookup
     assertThat(reverse.usages()).isEmpty();
   }
+
+  @Test
+  public void infer_fully_lambda_types() {
+    Result result = Result.createFor("InferLambdaType");
+    LambdaExpressionTree lambda = ((LambdaExpressionTree) result.symbol("line").declaration().parent());
+    JavaType lambdaType = (JavaType) lambda.symbolType();
+    assertThat(lambdaType.isParameterized()).isTrue();
+    assertThat(lambdaType.is("java.util.function.Function")).isTrue();
+    TypeSubstitution typeSubstitution = ((ParametrizedTypeJavaType) lambdaType).typeSubstitution;
+    assertThat(typeSubstitution.size()).isEqualTo(2);
+    JavaType Tsubstitution = typeSubstitution.substitutedTypes().get(0);
+    // check that T -> ? super String
+    assertThat(Tsubstitution.isTagged(JavaType.WILDCARD)).isTrue();
+    assertThat(((WildCardType) Tsubstitution).boundType).isEqualTo(WildCardType.BoundType.SUPER);
+    assertThat(((WildCardType) Tsubstitution).bound.is("java.lang.String")).isTrue();
+    // check that R -> String[]
```

Comment:
To me, we should end with `R -> ? extends String[]`.
It would also simplify the way to deduce what is the parameterization of the Stream returned by the map operation, as we would be able to compare two parameterization:

- `Function<? super String, ? extends String[]>` from our inference
- `Function<? super String, ? extends R>` from the signature of the map method: `<R> Stream<R> map(Function<? super T, ? extends R> mapper);`, already knowing `T` and searching for `R`.